adding files

joefitter opened the pull request:
This behaviour controls the address lookup process from postcode entry, to lookup and select, or manual entry if not found. The same step is used but the stage of the process you are on is controlled by a query param - `?step=postcode`. If omitted then `req.query.step` is set to postcode as default. `forks` and `next` as defined in step config work as expected, after the address lookup process has completed

joefitter committed:
adding files

JoeChapman commented on lib/defaults.js:
```diff
@@ -0,0 +1,10 @@
+    'not-found': 'Sorry – we couldn’t find any addresses for that postcode, enter your address manually',
+    'cant-connect': 'Sorry – we couldn’t connect to the postcode lookup service at this time, enter your address manually'
+  }
+}
```
could these be translations?

joefitter replied:
they are, these are fallbacks

joefitter replied:
if lookup fails (you forget or CBA to include a translation value) these are used as defaults. I think a standalone component/behaviour should have fallbacks for things like these to minimise setup in the majority of use cases

lennym replied:
I half agree and half disagree.

By failing early and showing the untranslated keys then there's an obvious cue to the user that they need to configure it properly, where these fallbacks are not really "proper".

However, I think it would be good to have some way of automating importing "default" translations for modules in the same way as we do with `hof-template-partials`. That would need some thought to not become unwieldy though.

JoeChapman commented:
Could you add a description, I'm not sure what it's supposed to do?

joefitter commented:
@JoeChapman have done. I don't want to be too specific as this is a first take and it may change following feedback from you guys

JoeChapman commented:
@joefitter thanks

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,261 @@
+    result = null;
+  }
+  return result;
+};
```
If this isn't already an option in `i18n-future` then I *really* need to do that.

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,261 @@
+
+const path = require('path');
+const querystring = require('querystring');
+const fetch = require('node-fetch');
```
We already have `hof-model` in the dependency stack which has all the functionality to do this. See https://github.com/UKHomeOffice/firearms/blob/master/apps/common/models/postcodes.js (EDIT: that's actually a pretty bad example because it doesn't use a bunch of the built-in convenience methods in the model - `auth` etc.)

I'd be wary of adding *another* way of talking to 3rd party APIs to the stack.

lennym replied:
I'm also not thrilled about baking the actual API fetch so hard into the behaviour as it couples us to one provider. I'd probably want to be able to pass a custom model into the behaviour that allows me to modify the fetching behaviour.

joefitter replied:
That is exactly how it was done before - with a model extending hof-model, it just felt like overkill when all we want to do is `fetch`

lennym replied:
If it weren't already guaranteed to be in the dependency tree already then I'd possibly agree, but it's already there, and encapsulating the functionality also allows for it to be more easily customisable by an end user.

joefitter replied:
I see what you mean - so give the user an option to pass in a model, but default to something like https://github.com/UKHomeOffice/end-tenancy/blob/master/apps/end-tenancy/models/postcodes.js? How would you deal with things like the auth code we currently set as a secret? Also the end tenancy implementation has a requirement that the address is in England, which involves a further api call.

lennym commented:
Immediate thoughts are that it looks *really* tightly coupled to one particular address provider. I don't know what that provider is, but it doesn't look like the one used in firearms.

I'd probably want to be able to re-use the journey but in a way which is configurable/extensible depending on API providers.

joefitter commented:
It is - the MOJ database which doesn't include Northern Ireland. It is used in _every_ SO service, including firearms. Reason being, it's free.

JoeChapman commented:
lol, who will break first

joefitter committed:
feature - split api logic to separate model

Added default model with fetch and validate logic. Added ability to override model to connect to different services

joefitter commented:
@lennym @JoeChapman @easternbloc @sulthan-ahmed - updated this now so it can be used with a custom Model if required

joefitter committed:
remove random thing

joefitter committed:
dont validate on api error

lennym reviewed (changes requested):
Needs some README.

joefitter commented:
@lennym absolutely, and some tests (which I am currently working on) - I wanted to get the teams take on it before adding a readme, in case it changes a lot

lennym commented:
At the very least I'd like some information about how it's used, and what its API looks like, even if it's not full documentation.

Just a "usage" section will help for now.

lennym commented on lib/default-model.js:
```diff
@@ -0,0 +1,73 @@
+        }
+        return resolve(data);
+      });
+    });
```
There's a lot of repetition of work that's already done in `hof-model#fetch` going on here. I'd look to set the minimum level of config and then delegate to `super`.

joefitter replied:
fair, this is mostly lifted from first implementation - I'll have a better look at hof-model

lennym replied:
I might have a look at supporting promises in `hof-model` too as that will reduce the need to extend a bit too.

joefitter replied:
either that or the implementation code could be changed to replace the promise with a callback

lennym commented on lib/default-model.js:
```diff
@@ -0,0 +1,73 @@
+      reqConf.method = 'GET';
+      reqConf.headers = {
+        Authorization: this.options.authorization
+      };
```
This is being duplicated from `fetch` where the model has an extensible `auth` method exactly for this purpose.

joefitter replied:
Looking at hof-model#auth now - thanks!

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,251 @@
+            `Code: ${err.status}; Detail: ${err.detail}`);
+          return callback();
+        });
+    }
```
Should we be calling `super.process` rather than `callback` on the happy path?

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,251 @@
+                }, req, res)
+              };
+            } else if (err.status === 403) {
+              return callback();
```
We'll want to call `super.validate` here in case other behaviours have been applied.

joefitter replied:
yep good spot 👍

joefitter committed:
* Use functions from hof-model for fetch, url and auth.
* Removed validateField and added custom validator
* Moved lookup to saveValues
* added README

joefitter committed:
call super from validate

joefitter committed:
Added functional testsm

joefitter commented:
@lennym changes made - not a clue why travis is failing...

lennym commented:
Maybe `rebase -i` away yesterday's commits?

joefitter commented:
@lennym squash merge will do that won't it?

lennym commented:
Still missing one instance of calling `callback` where we probably want to pass through to `super`. See my comment above on `process` method.

joefitter commented:
@lennym - I moved this to a separate method, its no longer in process so calling callback always calls super

lennym commented:
@joefitter They will still be referenced. Also I'd revert back to phantom rather than selenium for local use, which will happen magically if you rebase them out.

joefitter commented:
👍

joefitter commented:
@lennym rebased x

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,238 @@
+              };
+            } else if (err.status === 403 || err.status === 404) {
+              return super.validate(req, res, callback);
+            }
```
What's going on here with the status codes? The API returns 418 for invalid postcodes? 

It looks like we just happily continue on a 403/404, which doesn't seem quite correct.

lennym replied:
The code *looks* weird, so even if it's right I think it warrants a comment to explain why 403/404 are fine but 418 isn't.

joefitter replied:
you're right, one comment coming right up

joefitter replied:
@lennym added some comments. What do you think we should so if you _cant connect_ when validating? My thinking was that skipping api validation is all you can do in this situation. I guess if the validation is really important you could use a custom field validator as a backup. Filtering out all non english postcodes à la end tenancy would be a very nasty regex though

lennym replied:
Yeah, I guess since it's only a validation step it kinda makes sense just to plough on to lookup if it fails for some non-validation reason. Worst case scenario the lookup fails too and we show a message later.

:+1:

Thanks for humouring me with comments.

joefitter committed:
added comments to explain status codes
